gtsfm/utils/cycle_consistency.py

```diff
@@ -0,0 +1,233 @@
+    """Compute the cycle error by the magnitude of the axis-angle rotation after composing 3 rotations.
+
+    Note: i1 < i2 for every valid edge, by construction.
+
+    Args:
+        i2Ri1_dict: mapping from image pair indices to relative rotation.
+        cycle_nodes: 3-tuples of nodes that form a cycle. Nodes of are provided in sorted order.
+        two_view_reports_dict:
+        verbose: whether to dump to logger information about error in each Euler angle
+
+    Returns:
+        cycle_error: deviation from 3x3 identity matrix, in degrees. In other words,
+            it is defined as the magnitude of the axis-angle rotation of the composed transformations.
+        max_rot_error: maximum rotation error w.r.t. GT across triplet edges, in degrees.
+            If ground truth is not known for a scene, None will be returned instead.
+        max_trans_error: maximum translation error w.r.t. GT across triplet edges, in degrees.
+            If ground truth is not known for a scene, None will be returned instead.
+    """
+    cycle_nodes = list(cycle_nodes)
+    cycle_nodes.sort()
+
+    i0, i1, i2 = cycle_nodes
+
+    i1Ri0 = i2Ri1_dict[(i0, i1)]
+    i2Ri1 = i2Ri1_dict[(i1, i2)]
+    i0Ri2 = i2Ri1_dict[(i0, i2)].inverse()
+
+    # should compose to identity, with ideal measurements
+    i0Ri0 = i0Ri2.compose(i2Ri1).compose(i1Ri0)
+
+    I_3x3 = Rot3()
+    cycle_error = comp_utils.compute_relative_rotation_angle(I_3x3, i0Ri0)
+
+    # form 3 edges between fully connected subgraph (nodes i,j,k)
+    e_i = (i0, i1)
+    e_j = (i1, i2)
+    e_k = (i0, i2)
+
+    rot_errors = [two_view_reports_dict[e].R_error_deg for e in [e_i, e_j, e_k]]
+    trans_errors = [two_view_reports_dict[e].U_error_deg for e in [e_i, e_j, e_k]]
+
+    gt_known = all([err is not None for err in rot_errors])
+    if gt_known:
+        max_rot_error = np.max(rot_errors)
+        max_trans_error = np.max(trans_errors)
+    else:
+        # ground truth unknown, so cannot estimate error w.r.t. GT
+        max_rot_error = None
+        max_trans_error = None
+
+    if verbose:
+        i1Ri0_euler = Rotation.from_matrix(i1Ri0.matrix()).as_euler(seq="xyz", degrees=True).tolist()
+        i2Ri1_euler = Rotation.from_matrix(i2Ri1.matrix()).as_euler(seq="xyz", degrees=True).tolist()
+        i0Ri2_euler = Rotation.from_matrix(i0Ri2.matrix()).as_euler(seq="xyz", degrees=True).tolist()
+
+        euler_x = [i1Ri0_euler[0], i2Ri1_euler[0], i0Ri2_euler[0]]
+        euler_y = [i1Ri0_euler[1], i2Ri1_euler[1], i0Ri2_euler[1]]
+        euler_z = [i1Ri0_euler[2], i2Ri1_euler[2], i0Ri2_euler[2]]
+
+        logger.info("\n")
+        logger.info(f"{i0},{i1},{i2} --> Cycle error is: {cycle_error:.1f}")
+        if gt_known:
+            logger.info(f"Triplet: w/ max. R err {max_rot_error:.1f}, and w/ max. t err {max_trans_error:.1f}")
+
+        logger.info("X: (0->1) %.1f deg., (1->2) %.1f deg., (2->0) %.1f deg.", euler_x[0], euler_x[1], euler_x[2])
```

Comment:
why not directly use i1Ri0_euler[0], i2Ri1_euler[0], i0Ri2_euler[0] here?

Reply:
i thought this was a bit more readable, since otherwise the line gets quite wrong. but i'm fine either way

Reply:
I don’t quite like the idea of doing redundant work for logging. Also don’t think any readability is lost. You’ll save on the other lines even if these take up 2 lines. Also, is to_list() necessary?

Reply:
Good points. I removed it